### PR TITLE
Address EPA failure from issue 493

### LINKED ISSUE #493: EPA failure - another example of an EPA exception when computing signed distance
In the wild, another signed distance query produced an EPA failure. Here's the error message:

```
Original error message: external/fcl/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h:(1715): validateNearestFeatureOfPolytopeBeingEdge(): The origin is outside of the polytope. This should already have been identified as separating.
  Shape 1: Box0.5 0.5 0.5
  X_FS1
 0.96193976625564348026  0.15401279915737359216  0.22572537250328919556   2.0770063995786869349
-0.19134171618254486313  0.96936494951283913579  0.15401279915737359216  0.48468247475641951239
-0.19509032201612822033 -0.19134171618254486313  0.96193976625564348026   1.1543291419087275962
                      0                       0                       0                       1
  Shape 2: Box0.5 0.5 0.5
  X_FS2
 0.96193976625564348026  0.15401279915737359216  0.22572537250328919556                       2
-0.19134171618254486313  0.96936494951283913579  0.15401279915737359216                       0
-0.19509032201612822033 -0.19134171618254486313  0.96193976625564348026                    1.25
                      0                       0                       0                       1
  Solver: GjkSolver_libccd
    collision_tolerance:      2.0097183471152322134e-14
    max collision iterations: 500
    distance tolerance:       9.9999999999999995475e-07
    max distance iterations:  1000" thrown in the test body.
```

We need to reproduce this in test (see, e.g., [`test_distance_sphere_box_regression1()`](https://github.com/flexible-collision-library/fcl/blob/master/test/test_fcl_signed_distance.cpp#L467)).

Some quick observations:
  - Both boxes are the same size.
  - They have the same orientations.
  - It seems like they're stacked and canted slightly. It *seems* that they are in contact (so we would expect a small, negative signed distance), but that exact amount depends on the amount of canting. However, simply looking at the relative positions of the two transforms, its clearly on the order of 1 cm.

### PR DESCRIPTION
**Pull request by SeanCurtis-TRI**
Address EPA failure from issue 493

 - Incidentally, make the constants::eps* methods `constexpr`.
 - Introduce two regression tests (the two posted in the issue).
 - Correct a few confusing whitepsace formatting issues.
 - Modify the validateNearestFeatureOfPolytopeBeingEdge() function in two ways
   - Distance from origin to face must now be greater than epsilon (instead of zero). Justificaiton for this change provided.
   - Test to determine if the edge *truly* is the nearest feature has been simplified and documentation updated. This led to the removal of the is_point_on_line_segment() function.

The eps `constexpr` changes and the EPA fixes are in two separate commits. Let me know if you'd like them broken into two PRs.

resolves #493

---
This change is [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494)

### CODE REVIEW

**Review by SeanCurtis-TRI — COMMENTED** [review]
Note to reviewer:

R1: changes to `constants::fcl`.
R2: Fix for issue in #493 (which makes use of `constants::eps()`).

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MGsvkRD7YoenOv8mrMx:b-reb4t8)* status: 0 of 7 files reviewed, all discussions resolved

**Review by SeanCurtis-TRI — COMMENTED** [review]
Apparently `std::pow` is not `constexpr` in the clang available in CI. I've tweaked the `constants::eps` commit to accommodate this fact. Hopefully passing CI now.

R3 is part of R1. 

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MGtIG4q4sz0ovLwmGxd:b-5z9da6)* status: 0 of 6 files reviewed, all discussions resolved

**Review by SeanCurtis-TRI — COMMENTED** [review]
Looks like CI is happy. +@hongkai-dai willing to feature review?

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MGtbVtT2RZSwZYjLoSj:b-gljsfr)* status: 0 of 6 files reviewed, all discussions resolved (waiting on @hongkai-dai)

**Review by hongkai-dai — APPROVED** [review]
:lgtm:  +@sherm1 for platform review please, thanks!

> Reviewed 2 of 6 files at r1, 3 of 3 files at r2, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MGuNqS614kHj8gI-cI2:buohg44)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @sherm1)

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494)* status: all files reviewed, 1 unresolved discussion (waiting on @sherm1)

___
*[CHANGELOG.md, line 14 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-MGxWuLF1Yr2PJHpOydf:-MGxWuLF1Yr2PJHpOydg:b-nw7hlr) ([raw file](https://github.com/flexible-collision-library/fcl/blob/183c82f143170d88be7ce915e53e2e56046db3f0/CHANGELOG.md#L14)):*
> ```Markdown
> * Math
> 
>   * constants::eps (and variants) are now constexr:
> ```

nit: Forgot to update this when I removed the variants.

**Review by hongkai-dai — APPROVED** [review]
> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @sherm1)

**Review by sherm1 — APPROVED** [review]
Thanks, Sean! Platform :lgtm: with a few nits and a question.

> Reviewed 2 of 6 files at r1, 2 of 3 files at r2, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHCcyyYDM6GEhzgc6zI:b2emwpg)* status: all files reviewed, 3 unresolved discussions (waiting on @SeanCurtis-TRI)

___
*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1702 at r4](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-MHCc1wdACtualx2I2DY:-MHCc1wdACtualx2I2DZ:b-syl02y) ([raw file](https://github.com/flexible-collision-library/fcl/blob/2bb67a1b059efeb81547fa4d49822472c1d3ee23/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1702)):*
> ```objc
>     // We account for those differences by allowing a very small positive signed
>     // distance to be considered zero. We assume that the GJK/EPA code
>     // ultimately clasifies inside/outside around *zero* and *not* epsilon.
> ```

nit: clasifies -> classifies
___
*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1703 at r4](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-MHCcAY-0ImOyEkfBe_K:-MHCcAY-0ImOyEkfBe_L:b236jw5) ([raw file](https://github.com/flexible-collision-library/fcl/blob/2bb67a1b059efeb81547fa4d49822472c1d3ee23/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1703)):*
> ```objc
>     // distance to be considered zero. We assume that the GJK/EPA code
>     // ultimately clasifies inside/outside around *zero* and *not* epsilon.
>     if (origin_to_face_distance[i] > kEps) {
> ```

BTW this is a very tight tolerance (one roundoff bit). Is that enough?
___
*[test/test_fcl_signed_distance.cpp, line 487 at r4](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-MHCcnt9-JIoN49YIjLl:-MHCcnt9-JIoN49YIjLm:bsrp4qm) ([raw file](https://github.com/flexible-collision-library/fcl/blob/2bb67a1b059efeb81547fa4d49822472c1d3ee23/test/test_fcl_signed_distance.cpp#L487)):*
> ```cpp
>   X_WB2.translation() << 2, 0, 1.25;
>   const double expected_distance{0};
>   test_distance_box_box_helper(box_size, X_WB1, box_size, X_WB2, &expected_distance);
> ```

nit: long line
___
*[test/test_fcl_signed_distance.cpp, line 514 at r4](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-MHCctDF0oriwELydyet:-MHCctDF0oriwELydyeu:bsrp4qm) ([raw file](https://github.com/flexible-collision-library/fcl/blob/2bb67a1b059efeb81547fa4d49822472c1d3ee23/test/test_fcl_signed_distance.cpp#L514)):*
> ```cpp
>   X_WB2.translation() << 0, 0, 1.25;
>   const double expected_distance{0};
>   test_distance_box_box_helper(box_size, X_WB1, box_size, X_WB2, &expected_distance);
> ```

nit: long line

**Review by SeanCurtis-TRI — COMMENTED** [review]
Thanks for the review.

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHDFh5LARSIZTTaaKJ_:b-9qdoyq)* status: 4 of 6 files reviewed, all discussions resolved (waiting on @hongkai-dai and @sherm1)

___
*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1702 at r4](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-MHCc1wdACtualx2I2DY:-MHDBwpqCbsU9vPbNOj4:bb6un1j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/2bb67a1b059efeb81547fa4d49822472c1d3ee23/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1702)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

nit: clasifies -> classifies
</blockquote></details>

Done
___
*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1703 at r4](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-MHCcAY-0ImOyEkfBe_K:-MHDByNt3WUlQRjHlOgT:b-xvelx1) ([raw file](https://github.com/flexible-collision-library/fcl/blob/2bb67a1b059efeb81547fa4d49822472c1d3ee23/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1703)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW this is a very tight tolerance (one roundoff bit). Is that enough?
</blockquote></details>

Done

So, there are two issues here:

1. Is one bit enough for properly scaled problems?
2. What about badly scaled problems?

I directly addressed item 2. As for item 1, I'm going to suggest that all the tests pass, and that's a pretty good starting point. I'm loathe to dial it up without some specific rationale (theoretical or empirical).
___
*[test/test_fcl_signed_distance.cpp, line 487 at r4](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-MHCcnt9-JIoN49YIjLl:-MHDCODe6JNO2VCKezrZ:bb6un1j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/2bb67a1b059efeb81547fa4d49822472c1d3ee23/test/test_fcl_signed_distance.cpp#L487)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

nit: long line
</blockquote></details>

Done
___
*[test/test_fcl_signed_distance.cpp, line 514 at r4](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-MHCctDF0oriwELydyet:-MHDCQf-0PiypbnfQVng:bb6un1j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/2bb67a1b059efeb81547fa4d49822472c1d3ee23/test/test_fcl_signed_distance.cpp#L514)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

nit: long line
</blockquote></details>

Done

**Review by sherm1 — APPROVED** [review]
> Reviewed 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by sherm1 — APPROVED** [review]
Ready to merge -- I'll leave that to you @seancurtis-tri in case you want to revise the long commit message upon squashing.

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHDZul5D4R-CqsynBg4:bf0k6of)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by SeanCurtis-TRI — COMMENTED** [review]
Thanks. I'm running it through "other CI" to make sure that nothing is obviously hurt. Once that gets through, I'll merge here.

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHDlmqj8oMM_Psmx3G2:b6ljfup)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by SeanCurtis-TRI — COMMENTED** [review]
Reviewers, I decided to update the test prior to merging based on changes due to reviewer comments and an additional failure case documented in the issue. PTAL.

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHHT9fK66d4JZW1oeVW:bc2bx7p)* status: 5 of 6 files reviewed, all discussions resolved (waiting on @sherm1)

**Review by sherm1 — APPROVED** [review]
:lgtm: Just to make sure -- did you verify that the new more-generic set of tests fails without the fix here?

> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHHXp9s-5aRGm2aze6k:b334jdr)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by SeanCurtis-TRI — COMMENTED** [review]
Of course.  :) I pulled the test onto master and confirmed failure. 

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHHYQr806HUTGK1PAqw:b-4kqq6a)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by SeanCurtis-TRI — COMMENTED** [review]
Blech...mac failure.

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHHvSJO-GmfzRobUYO-:b7dnfe0)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by sherm1 — APPROVED** [review]
:angry:

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHHx7DLAVKiQaTNIWVB:bb3n0su)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by SeanCurtis-TRI — COMMENTED** [review]
The mac failure was due to one particular case where the error was ever so slightly larger than epsilon. So, I bumped the value of epsilon by 1 bit and it made it happy on macsim. Let's see if that makes CI happy.

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHI14PBEsrCFqZxyO-h:b-9v417x)* status: 5 of 6 files reviewed, all discussions resolved (waiting on @sherm1)

**Review by SeanCurtis-TRI — COMMENTED** [review]
@sherm1 Can you take a look at the latest revision to address mac CI failure and let me know if you buy it?

> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHIYeIGCqaLEmsxb0lU:b2a314e)* status: 5 of 6 files reviewed, all discussions resolved (waiting on @sherm1)

**Review by sherm1 — APPROVED** [review]
I'm happy with that, Sean. The 1-bit eps was making me nervous anyway -- that's a lot to ask out of a numerical calculation!

> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/494#-:-MHImj8cE_F1DwbWqvL_:b90r96s)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
